File: 215. Kth Largest Element in an Array.cpp
// https://leetcode.com/problems/kth-largest-element-in-an-array/

// Given an integer array nums and an integer k, return the kth largest element in
// the array.

// Note that it is the kth largest element in the sorted order, not the kth
// distinct element.

////////////////////////////////////////////////////////////////////////////////

// quick selection, sort(nums)[n-k]
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        k = nums.size() - k; // find sort(nums)[n-k]
        quickSelect(nums, 0, nums.size() - 1, k);
        return nums[k];
    }
private:    
    int partition(vector<int>& nums, int l, int r, int p) {
        int pVal = nums[p]; swap(nums[p], nums[r]);
        int anchor = l;
        for (int i = l; i < r; ++i) {
            if (nums[i] < pVal) swap(nums[anchor++], nums[i]);
        }
        swap(nums[anchor], nums[r]);
        return anchor;
    }
    void quickSelect(vector<int>& nums, int l, int r, int k) {
        int m = l + (r - l) / 2; // use m as pivot
        int p = partition(nums, l, r, m);
        if (p == k) return;
        if (p > k) quickSelect(nums, l, p - 1, k);
        else quickSelect(nums, p + 1, r, k);
        return;
    }
};

File: 215. Kth Largest Element in an Array.py
# https://leetcode.com/problems/kth-largest-element-in-an-array/

# Given an integer array nums and an integer k, return the kth largest element in
# the array.

# Note that it is the kth largest element in the sorted order, not the kth
# distinct element.

################################################################################

# find nums.sort()[n-k]
# use quick_select

class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        k = len(nums) - k # find nums.sort()[n-k]
        self.quick_select(nums, 0, len(nums) - 1, k)
        return nums[k]

    def quick_select(self, nums, start, end, k):
        mid = start + (end - start) // 2
        piv = self.partition(nums, start, end, mid)
        if piv == k:
            return
        elif piv < k:
            self.quick_select(nums, piv + 1, end, k)
        else:
            self.quick_select(nums, start, piv - 1, k)
        return

    def partition(self, nums, start, end, piv):
        piv_val = nums[piv]
        # put piv to end
        nums[piv], nums[end] = nums[end], nums[piv]
        # put smaller number first
        store_idx = start
        for i in range(start, end):
            if nums[i] < piv_val:
                nums[i], nums[store_idx] = nums[store_idx], nums[i]
                store_idx += 1
        # restore piv
        nums[end], nums[store_idx] = nums[store_idx], nums[end]
        return store_idx 
